src/mon/commands/Command.h

```diff
@@ -0,0 +1,330 @@
+
+    string format;
+    cmd_getval(cct, cmdmap, "format", format, string("plain"));
+    FormatterRef f(Formatter::create(format));
+
+    string prefix;
+    cmd_getval(cct, cmdmap, "prefix", prefix);
+
+    return do_prepare(op, prefix, cmdmap,
+	ss, rdata, f, args...);
+  }
+
+  virtual bool do_preprocess(
+      MonOpRequestRef op,
+      const string &prefix,
+      const cmdmap_t &cmdmap,
+      stringstream &ss,
+      bufferlist rdata,
+      FormatterRef f,
+      Args ...args) = 0;
+
+  virtual bool do_prepare(
+      MonOpRequestRef op,
+      const string &prefix,
+      const cmdmap_t &cmdmap,
+      stringstream &ss,
+      bufferlist rdata,
+      FormatterRef f,
+      Args ...args) = 0;
+
+  virtual bool handles_command(const string &prefix) = 0;
+
+ private:
+  Command() = delete;
+};
+
+
+/* ReadCommand
+ *
+ * Defines a templated Read-only command. These will be called during
+ * preprocess. This class abstracts a derived class' command definition, by
+ * redefining the preprocess command from the 'Command' class, and then
+ * calling a function to be implemented by the class deriving from this one.
+ *
+ * Preparing is still left defined as the 'Command' class has it defined,
+ * which is basically returning 'false'.
+ */
+template<typename T, typename ...Args>
+struct ReadCommand : public Command<T, Args...>
+{
+
+  explicit ReadCommand(
+      Monitor *_mon,
+      T *_svc,
+      CephContext *_cct) :
+    Command<T, Args...>(_mon, _svc, _cct)
+  { };
+  virtual ~ReadCommand() { }
+
+  using Command<T, Args...>::prepare;
+  virtual bool prepare(
+      MonOpRequestRef op,
+      const cmdmap_t &cmdmap,
+      Args ...args) final {
+    ceph_assert(0 == "read-only commands do not perform prepare phase");
+    return false;
+  }
+
+  virtual bool do_prepare(
+      MonOpRequestRef op,
+      const string &prefix,
+      const cmdmap_t &cmdmap,
+      stringstream &ss,
+      bufferlist rdata,
+      FormatterRef f,
+      Args ...args) final {
+    ceph_assert(0 == "read-only commands do not perform prepare phase");
+    return false;
+  }
+
+ private:
+  ReadCommand() = delete;
+
+};
+
+/* WriteCommand
+ *
+ * Defines a templated a Write command. These will be called during prepare.
+ * This class abstracts a derived class' command definition for the prepare
+ * stage, redefining prepare function inherited from the 'Command' class.
+ *
+ * For commands that will have both the preprocess and prepare stages, said
+ * command implementation will require to derive from both the 'ReadCommand'
+ * and 'WriteCommand' classes, and implement both 'do_{preprocess,prepare}'
+ * functions.
+ */
+template<typename T, typename ...Args>
+struct WriteCommand : public Command<T, Args...>
+{
+
+  explicit WriteCommand(
+      Monitor *_mon,
+      T *_svc,
+      CephContext *_cct) :
+    Command<T, Args...>(_mon, _svc, _cct)
+  { }
+  virtual ~WriteCommand() { }
+
+  using Command<T, Args...>::preprocess;
+  virtual bool preprocess(
```

Comment:
Consider marking this "final", as I think that will catch such uses at compile time.